### ISSUE TITLE
Missing dll in NetFabric.Hyperlinq 3.0.0-beta12

### ISSUE THREAD

**Issue by gregcotten** [NetFabric/NetFabric.Hyperlinq#161]
On NuGet, [NetFabric.Hyperlinq](https://www.nuget.org/packages/NetFabric.Hyperlinq/3.0.0-beta12) does not have a .dll in the lib folder. Is this correct?

**Comment by gregcotten** [comment]
I am using the [NuGetForUnity](https://github.com/GlitchEnzo/NuGetForUnity) package manager, but I don't think that is the culprit!

**Comment by aalmada** [comment]
Hi @gregcotten 
The package contains DLLs for targets `netcoreapp2.1` and `netstandard2.1`. Unfortunately, Unity does not yet support any of these. 
Sorry, but the handling of older targets was becoming too time-consuming.
I just hope Unity will support .NET 5.0 as soon as possible.

**Comment by gregcotten** [comment]
Oh interesting - it seems .NET 5.0 isn't actually released until November 2020, so they probably won't support it until its official release. What's the latest version of Hyperlinq that supports .NET 4.x?

**Comment by wilg** [comment]
@aalmada Can you give any pointers on how we might be able to get things to build on `.NET Standard 2.0`. This is the version Unity currently supports and I suspect they won't add support for a while.

Non-allocating Linq is crucial for Unity development because per-frame allocations kill game performance, so this library would be absolutely perfect.

Is it possible that just compiling a version for netstandard2.0 could happen with a subset of features, such as without Span/ReadOnlySpan? We really just need the basics, and it's surprisingly hard to find a solution for Unity!

Happy to do some of the work, but some assistance from someone with more C#/.NET experience would be amazing. Thanks!

**Comment by aalmada** [comment]
@gregcotten @wilg 

I don't often develop on Unity but I strongly believe that it should be supported. That was my initial plan but managing the multiple versions of the code became a burden and I haven't seen much interest expressed here. So, thank you so much for your feedback.

I use value-types as much as possible so that there are no heap allocations. This means I cannot use inheritance to reuse code. I have to copy and adapt to each implementation. Every time I find a better solution, I have to go through every file and make the changes. I'm using code generators to automate some of this but there's still too much manual labor.

Lately, I've been investing my time only on increasing the code coverage and fixing the issues found so that I can really call this a stable version. 

I was now thinking that, given that I have so many tests, the reasons that made me drop Unity support are not as strong. I think I can handle the arrays as `IReadOnlyList<>`, for  `netstandard2.0`, and `ReadOnlySpan<>` for other target platforms.

Thanks for the offer to help. Let me see what I can do.

**Comment by wilg** [comment]
That sounds amazing! Thanks! Excited to hear how it goes and if I can be any help.

**Comment by aalmada** [comment]
Ended up being a lot easier than I expected. I already merged #165 and released [3.0.0-beta13](https://www.nuget.org/packages/NetFabric.Hyperlinq/3.0.0-beta13). Thanks for making me look it to this again.

Please let me know if you find any issues.

@wilg Thank you so much for coveralls. It's been very useful in the development of this project.

**Comment by wilg** [comment]
Awesome! Thanks so much for doing that! We'll check it out soon!

Really cool you're using Coveralls, haven't been a part of it in a while but always awesome to see people using stuff you made!

**Comment by aalmada** [comment]
I run some benchmarks to understand the performance on all target platforms. 

For `netstandard2.0`, the easiest to maintain solution for arrays is to handle them as `IReadOnlyList` but the performance is not great, although I use the indexer instead of the enumerator. I'll have to look into that.

In case you want to go through the results. The benchmark runs the operation `source.Where(item => (item & 0x01) == 0)`, which returns the even values. `source` is a collection of 100 items, with the values from 0 to 99. 

The name of the method gives you the library used (LINQ, [LinqFaster](https://github.com/jackmott/LinqFaster) or Hyperlinq), the type of collection (array, `IEnumerable<>`, `IReadOnlyCollection<>` or `IReadOnlyList`) and the type of enumerator (value-type or reference-type). A boxed value-type enumerator becomes a reference-type enumerator. This happens when the collection is cast to an interface. To help you spot these cases, I suggest the use of the analyzer [NetFabric.Hyperlinq.Analyzer](https://github.com/NetFabric/NetFabric.Hyperlinq.Analyzer).

The 'Runtime' column, gives you the .NET runtime used (`net461`, `netcoreapp2.1` or `netcoreapp3.1).

Looking at the 'Ratio' column, you can see that Hyperlinq performs better than LINQ, except for arrays on `net461`.

Looking at the `Allocated` column, you can see that Hypelinq does not allocate on the heap, except for reference-type enumerators but, still allocates less than LINQ. 

[LinqFaster](https://github.com/jackmott/LinqFaster) is faster on arrays but allocates much more memory.

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.264 (2004/?/20H1)
Intel Core i7-7567U CPU 3.50GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
  [Host]        : .NET Framework 4.8 (4.8.4180.0), X64 RyuJIT
  .NET 4.6.1    : .NET Framework 4.8 (4.8.4180.0), X64 RyuJIT
  .NET Core 2.1 : .NET Core 2.1.16 (CoreCLR 4.6.28516.03, CoreFX 4.6.28516.10), X64 RyuJIT
  .NET Core 3.1 : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT


```
|                              Method |           Job |       Runtime |                Categories | Count |        Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------ |-------------- |-------------- |-------------------------- |------ |------------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
|                          Linq_Array |    .NET 4.6.1 |    .NET 4.6.1 |                     Array |   100 |    508.6 ns |  8.56 ns | 11.13 ns |  1.00 |    0.00 | 0.0229 |     - |     - |      48 B |
|                    LinqFaster_Array |    .NET 4.6.1 |    .NET 4.6.1 |                     Array |   100 |    328.1 ns |  1.19 ns |  0.99 ns |  0.64 |    0.01 | 0.3095 |     - |     - |     650 B |
|                     Hyperlinq_Array |    .NET 4.6.1 |    .NET 4.6.1 |                     Array |   100 |    729.1 ns |  5.61 ns |  5.25 ns |  1.42 |    0.03 |      - |     - |     - |         - |
|                          Linq_Array | .NET Core 2.1 | .NET Core 2.1 |                     Array |   100 |    481.3 ns |  2.39 ns |  1.99 ns |  0.94 |    0.02 | 0.0219 |     - |     - |      48 B |
|                    LinqFaster_Array | .NET Core 2.1 | .NET Core 2.1 |                     Array |   100 |    346.2 ns |  3.61 ns |  3.38 ns |  0.68 |    0.02 | 0.3085 |     - |     - |     648 B |
|                     Hyperlinq_Array | .NET Core 2.1 | .NET Core 2.1 |                     Array |   100 |    390.4 ns |  1.57 ns |  1.31 ns |  0.77 |    0.02 |      - |     - |     - |         - |
|                          Linq_Array | .NET Core 3.1 | .NET Core 3.1 |                     Array |   100 |    422.3 ns |  1.99 ns |  1.87 ns |  0.82 |    0.02 | 0.0229 |     - |     - |      48 B |
|                    LinqFaster_Array | .NET Core 3.1 | .NET Core 3.1 |                     Array |   100 |    346.0 ns |  1.48 ns |  1.23 ns |  0.68 |    0.02 | 0.3095 |     - |     - |     648 B |
|                     Hyperlinq_Array | .NET Core 3.1 | .NET Core 3.1 |                     Array |   100 |    365.7 ns |  1.21 ns |  0.95 ns |  0.72 |    0.02 |      - |     - |     - |         - |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|               Linq_Enumerable_Value |    .NET 4.6.1 |    .NET 4.6.1 |          Enumerable_Value |   100 |  1,038.4 ns | 10.51 ns |  8.77 ns |  1.00 |    0.00 | 0.0496 |     - |     - |     104 B |
|          Hyperlinq_Enumerable_Value |    .NET 4.6.1 |    .NET 4.6.1 |          Enumerable_Value |   100 |    354.2 ns |  1.42 ns |  1.26 ns |  0.34 |    0.00 |      - |     - |     - |         - |
|               Linq_Enumerable_Value | .NET Core 2.1 | .NET Core 2.1 |          Enumerable_Value |   100 |  1,130.7 ns |  9.05 ns |  8.02 ns |  1.09 |    0.01 | 0.0477 |     - |     - |     104 B |
|          Hyperlinq_Enumerable_Value | .NET Core 2.1 | .NET Core 2.1 |          Enumerable_Value |   100 |    351.8 ns |  1.95 ns |  1.52 ns |  0.34 |    0.00 |      - |     - |     - |         - |
|               Linq_Enumerable_Value | .NET Core 3.1 | .NET Core 3.1 |          Enumerable_Value |   100 |  1,116.4 ns |  5.34 ns |  4.17 ns |  1.08 |    0.01 | 0.0496 |     - |     - |     104 B |
|          Hyperlinq_Enumerable_Value | .NET Core 3.1 | .NET Core 3.1 |          Enumerable_Value |   100 |    344.9 ns |  3.52 ns |  3.12 ns |  0.33 |    0.00 |      - |     - |     - |         - |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|               Linq_Collection_Value |    .NET 4.6.1 |    .NET 4.6.1 |          Collection_Value |   100 |  1,065.9 ns | 11.80 ns |  9.86 ns |  1.00 |    0.00 | 0.0496 |     - |     - |     104 B |
|          Hyperlinq_Collection_Value |    .NET 4.6.1 |    .NET 4.6.1 |          Collection_Value |   100 |    381.5 ns |  1.25 ns |  1.11 ns |  0.36 |    0.00 |      - |     - |     - |         - |
|               Linq_Collection_Value | .NET Core 2.1 | .NET Core 2.1 |          Collection_Value |   100 |  1,131.1 ns |  8.50 ns |  7.10 ns |  1.06 |    0.01 | 0.0477 |     - |     - |     104 B |
|          Hyperlinq_Collection_Value | .NET Core 2.1 | .NET Core 2.1 |          Collection_Value |   100 |    367.9 ns |  1.52 ns |  1.27 ns |  0.35 |    0.00 |      - |     - |     - |         - |
|               Linq_Collection_Value | .NET Core 3.1 | .NET Core 3.1 |          Collection_Value |   100 |  1,096.3 ns |  7.71 ns |  6.44 ns |  1.03 |    0.01 | 0.0496 |     - |     - |     104 B |
|          Hyperlinq_Collection_Value | .NET Core 3.1 | .NET Core 3.1 |          Collection_Value |   100 |    343.1 ns |  1.48 ns |  1.31 ns |  0.32 |    0.00 |      - |     - |     - |         - |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|                     Linq_List_Value |    .NET 4.6.1 |    .NET 4.6.1 |                List_Value |   100 |  1,063.6 ns |  5.93 ns |  5.55 ns |  1.00 |    0.00 | 0.0496 |     - |     - |     104 B |
|                Hyperlinq_List_Value |    .NET 4.6.1 |    .NET 4.6.1 |                List_Value |   100 |    377.7 ns |  1.34 ns |  1.19 ns |  0.35 |    0.00 |      - |     - |     - |         - |
|                     Linq_List_Value | .NET Core 2.1 | .NET Core 2.1 |                List_Value |   100 |  1,115.2 ns |  9.14 ns |  7.64 ns |  1.05 |    0.01 | 0.0477 |     - |     - |     104 B |
|                Hyperlinq_List_Value | .NET Core 2.1 | .NET Core 2.1 |                List_Value |   100 |    368.1 ns |  1.40 ns |  1.31 ns |  0.35 |    0.00 |      - |     - |     - |         - |
|                     Linq_List_Value | .NET Core 3.1 | .NET Core 3.1 |                List_Value |   100 |  1,073.9 ns |  7.83 ns |  6.54 ns |  1.01 |    0.01 | 0.0496 |     - |     - |     104 B |
|                Hyperlinq_List_Value | .NET Core 3.1 | .NET Core 3.1 |                List_Value |   100 |    370.8 ns |  1.44 ns |  1.12 ns |  0.35 |    0.00 |      - |     - |     - |         - |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|          Linq_AsyncEnumerable_Value |    .NET 4.6.1 |    .NET 4.6.1 |     AsyncEnumerable_Value |   100 |  9,962.4 ns | 43.72 ns | 34.13 ns |  1.00 |    0.00 | 0.0458 |     - |     - |     120 B |
|     Hyperlinq_AsyncEnumerable_Value |    .NET 4.6.1 |    .NET 4.6.1 |     AsyncEnumerable_Value |   100 |  7,262.9 ns | 69.24 ns | 64.77 ns |  0.73 |    0.01 |      - |     - |     - |         - |
|          Linq_AsyncEnumerable_Value | .NET Core 2.1 | .NET Core 2.1 |     AsyncEnumerable_Value |   100 |  6,895.4 ns | 22.18 ns | 19.66 ns |  0.69 |    0.00 | 0.0458 |     - |     - |     120 B |
|     Hyperlinq_AsyncEnumerable_Value | .NET Core 2.1 | .NET Core 2.1 |     AsyncEnumerable_Value |   100 |  5,633.0 ns | 19.04 ns | 15.90 ns |  0.57 |    0.00 |      - |     - |     - |         - |
|          Linq_AsyncEnumerable_Value | .NET Core 3.1 | .NET Core 3.1 |     AsyncEnumerable_Value |   100 |  6,781.3 ns | 34.61 ns | 32.37 ns |  0.68 |    0.00 | 0.0534 |     - |     - |     120 B |
|     Hyperlinq_AsyncEnumerable_Value | .NET Core 3.1 | .NET Core 3.1 |     AsyncEnumerable_Value |   100 |  5,267.0 ns | 23.79 ns | 21.09 ns |  0.53 |    0.00 |      - |     - |     - |         - |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|           Linq_Enumerable_Reference |    .NET 4.6.1 |    .NET 4.6.1 |      Enumerable_Reference |   100 |    918.2 ns |  9.76 ns |  9.13 ns |  1.00 |    0.00 | 0.0458 |     - |     - |      96 B |
|      Hyperlinq_Enumerable_Reference |    .NET 4.6.1 |    .NET 4.6.1 |      Enumerable_Reference |   100 |    885.5 ns |  4.28 ns |  3.57 ns |  0.97 |    0.01 | 0.0191 |     - |     - |      40 B |
|           Linq_Enumerable_Reference | .NET Core 2.1 | .NET Core 2.1 |      Enumerable_Reference |   100 |  1,026.5 ns |  9.27 ns |  7.74 ns |  1.12 |    0.01 | 0.0439 |     - |     - |      96 B |
|      Hyperlinq_Enumerable_Reference | .NET Core 2.1 | .NET Core 2.1 |      Enumerable_Reference |   100 |    916.7 ns |  7.75 ns |  7.25 ns |  1.00 |    0.01 | 0.0181 |     - |     - |      40 B |
|           Linq_Enumerable_Reference | .NET Core 3.1 | .NET Core 3.1 |      Enumerable_Reference |   100 |    950.8 ns |  6.65 ns |  6.22 ns |  1.04 |    0.01 | 0.0458 |     - |     - |      96 B |
|      Hyperlinq_Enumerable_Reference | .NET Core 3.1 | .NET Core 3.1 |      Enumerable_Reference |   100 |    919.3 ns |  3.14 ns |  2.94 ns |  1.00 |    0.01 | 0.0191 |     - |     - |      40 B |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|           Linq_Collection_Reference |    .NET 4.6.1 |    .NET 4.6.1 |      Collection_Reference |   100 |    847.0 ns |  3.71 ns |  3.29 ns |  1.00 |    0.00 | 0.0381 |     - |     - |      80 B |
|      Hyperlinq_Collection_Reference |    .NET 4.6.1 |    .NET 4.6.1 |      Collection_Reference |   100 |    830.5 ns |  3.66 ns |  2.86 ns |  0.98 |    0.00 | 0.0114 |     - |     - |      24 B |
|           Linq_Collection_Reference | .NET Core 2.1 | .NET Core 2.1 |      Collection_Reference |   100 |    948.3 ns |  2.77 ns |  2.45 ns |  1.12 |    0.00 | 0.0372 |     - |     - |      80 B |
|      Hyperlinq_Collection_Reference | .NET Core 2.1 | .NET Core 2.1 |      Collection_Reference |   100 |    838.3 ns |  5.14 ns |  4.55 ns |  0.99 |    0.01 | 0.0105 |     - |     - |      24 B |
|           Linq_Collection_Reference | .NET Core 3.1 | .NET Core 3.1 |      Collection_Reference |   100 |    867.3 ns |  2.71 ns |  2.53 ns |  1.02 |    0.01 | 0.0381 |     - |     - |      80 B |
|      Hyperlinq_Collection_Reference | .NET Core 3.1 | .NET Core 3.1 |      Collection_Reference |   100 |    844.9 ns |  6.72 ns |  5.95 ns |  1.00 |    0.01 | 0.0114 |     - |     - |      24 B |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|                 Linq_List_Reference |    .NET 4.6.1 |    .NET 4.6.1 |            List_Reference |   100 |    796.1 ns |  3.38 ns |  2.82 ns |  1.00 |    0.00 | 0.0381 |     - |     - |      80 B |
|            Hyperlinq_List_Reference |    .NET 4.6.1 |    .NET 4.6.1 |            List_Reference |   100 |    645.6 ns |  3.31 ns |  2.93 ns |  0.81 |    0.00 |      - |     - |     - |         - |
|                 Linq_List_Reference | .NET Core 2.1 | .NET Core 2.1 |            List_Reference |   100 |    952.5 ns |  5.27 ns |  4.93 ns |  1.20 |    0.01 | 0.0362 |     - |     - |      80 B |
|            Hyperlinq_List_Reference | .NET Core 2.1 | .NET Core 2.1 |            List_Reference |   100 |    648.9 ns |  2.53 ns |  1.98 ns |  0.82 |    0.00 |      - |     - |     - |         - |
|                 Linq_List_Reference | .NET Core 3.1 | .NET Core 3.1 |            List_Reference |   100 |    919.2 ns |  4.31 ns |  3.60 ns |  1.15 |    0.00 | 0.0381 |     - |     - |      80 B |
|            Hyperlinq_List_Reference | .NET Core 3.1 | .NET Core 3.1 |            List_Reference |   100 |    613.9 ns |  3.59 ns |  3.36 ns |  0.77 |    0.01 |      - |     - |     - |         - |
|                                     |               |               |                           |       |             |          |          |       |         |        |       |       |           |
|      Linq_AsyncEnumerable_Reference |    .NET 4.6.1 |    .NET 4.6.1 | AsyncEnumerable_Reference |   100 | 15,388.3 ns | 80.20 ns | 66.97 ns |  1.00 |    0.00 | 0.0305 |     - |     - |      64 B |
| Hyperlinq_AsyncEnumerable_Reference |    .NET 4.6.1 |    .NET 4.6.1 | AsyncEnumerable_Reference |   100 | 12,875.8 ns | 44.14 ns | 36.86 ns |  0.84 |    0.00 |      - |     - |     - |         - |
|      Linq_AsyncEnumerable_Reference | .NET Core 2.1 | .NET Core 2.1 | AsyncEnumerable_Reference |   100 |  9,330.7 ns | 38.47 ns | 30.04 ns |  0.61 |    0.00 | 0.0153 |     - |     - |      64 B |
| Hyperlinq_AsyncEnumerable_Reference | .NET Core 2.1 | .NET Core 2.1 | AsyncEnumerable_Reference |   100 |  8,434.2 ns | 39.08 ns | 32.63 ns |  0.55 |    0.00 |      - |     - |     - |         - |
|      Linq_AsyncEnumerable_Reference | .NET Core 3.1 | .NET Core 3.1 | AsyncEnumerable_Reference |   100 |  8,582.6 ns | 56.91 ns | 50.45 ns |  0.56 |    0.00 | 0.0305 |     - |     - |      64 B |
| Hyperlinq_AsyncEnumerable_Reference | .NET Core 3.1 | .NET Core 3.1 | AsyncEnumerable_Reference |   100 |  8,251.6 ns | 43.80 ns | 40.97 ns |  0.54 |    0.00 |      - |     - |     - |         - |

**Comment by aalmada** [comment]
#167 fixes the performance issues with operations on arrays in `standard2.0`. Released as [3.0.0-beta14](https://www.nuget.org/packages/NetFabric.Hyperlinq/3.0.0-beta14).
